core/src/main/java/org/apache/iceberg/avro/AvroMetrics.java

```diff
@@ -31,7 +42,112 @@ private AvroMetrics() {
 
   static Metrics fromWriter(DatumWriter<?> datumWriter, Schema schema, long numRecords,
                             MetricsConfig inputMetricsConfig) {
-    // TODO will populate in following PRs if datum writer is a MetricsAwareDatumWriter
-    return new Metrics(numRecords, null, null, null);
+    if (!(datumWriter instanceof MetricsAwareDatumWriter)) {
+      return new Metrics(numRecords, null, null, null, null);
+    }
+
+    MetricsAwareDatumWriter<?> metricsAwareDatumWriter = (MetricsAwareDatumWriter<?>) datumWriter;
+    MetricsConfig metricsConfig;
+    if (inputMetricsConfig == null) {
+      metricsConfig = MetricsConfig.getDefault();
+    } else {
+      metricsConfig = inputMetricsConfig;
+    }
+
+    Map<Integer, Long> valueCounts = new HashMap<>();
+    Map<Integer, Long> nullValueCounts = new HashMap<>();
+    Map<Integer, Long> nanValueCounts = new HashMap<>();
+    Map<Integer, ByteBuffer> lowerBounds = new HashMap<>();
+    Map<Integer, ByteBuffer> upperBounds = new HashMap<>();
```

Comment:
One thing to consider is that quite a bit of this method and the helper methods below is generic and could be written for `Stream<FieldMetrics>`. I don't think it needs to happen right now, but I think it would be good to have this separated into the Avro-specific part and a metrics part that lives in `MetricsUtil`.